exchanges/orderbook/incremental_updates.go

```diff
@@ -0,0 +1,234 @@
+
+	if d.validationError != nil {
+		return d.validationError
+	}
+
+	// This will process out of order updates but will not error on them.
+	// TODO: Error on out of order updates; this is intentionally kept as is from the buffer package.
+	// Add update.UpdateTime time check to ensure that the update is newer than the last update,
+	// this should screen zero values as well.
+	if u.SkipOutOfOrderLastUpdateID && d.lastUpdateID >= u.UpdateID {
+		return nil
+	}
+
+	if d.options.restSnapshot {
+		return d.invalidate(errRESTSnapshot)
+	}
+
+	if u.Action != 0 {
+		if err := d.updateByIDAndAction(u); err != nil {
+			return d.invalidate(err)
+		}
+	} else {
+		if err := d.updateBidAskByPrice(u); err != nil {
+			return d.invalidate(err)
+		}
+	}
+
+	if u.ExpectedChecksum != 0 {
+		if u.GenerateChecksum == nil {
+			return d.invalidate(errChecksumGeneratorUnset)
+		}
+		if checksum := u.GenerateChecksum(d.snapshot()); checksum != u.ExpectedChecksum {
+			return d.invalidate(fmt.Errorf("%s %s %s %w: expected '%d', got '%d'", d.exchange, d.pair, d.asset, errChecksumMismatch, u.ExpectedChecksum, checksum))
+		}
+	} else if d.verifyOrderbook {
+		if err := verify(d.snapshot()); err != nil {
+			return d.invalidate(err)
+		}
+	}
+
+	return nil
+}
+
+// // TODO: Confirm no alloc as it shouldn't escape
+func (d *Depth) snapshot() *Base {
+	return &Base{
+		Bids:                   d.bidTranches.Tranches,
+		Asks:                   d.askTranches.Tranches,
+		Exchange:               d.options.exchange,
+		Pair:                   d.pair,
+		Asset:                  d.asset,
+		IsFundingRate:          d.options.isFundingRate,
+		PriceDuplication:       d.options.priceDuplication,
+		IDAlignment:            d.options.idAligned,
+		ChecksumStringRequired: d.options.checksumStringRequired,
+	}
+}
+
+// updateByIDAndAction will receive an action to execute against the orderbook it will then match by IDs instead of
+// price to perform the action
+func (d *Depth) updateByIDAndAction(u *Update) error {
+	switch u.Action {
+	case Amend:
+		if err := d.updateBidAskByID(u); err != nil {
+			return fmt.Errorf("%w %w", errAmendFailure, err)
+		}
+	case Delete:
+		// edge case for Bitfinex as their streaming endpoint duplicates deletes
+		bypassErr := d.options.exchange == "Bitfinex" && d.options.isFundingRate // TODO: Confirm this is still correct
+		if err := d.deleteBidAskByID(u, bypassErr); err != nil {
+			return fmt.Errorf("%w %w", errDeleteFailure, err)
+		}
+	case Insert:
+		if err := d.insertBidAskByID(u); err != nil {
+			return fmt.Errorf("%w %w", errInsertFailure, err)
+		}
+	case UpdateInsert:
+		if err := d.updateInsertByID(u); err != nil {
+			return fmt.Errorf("%w %w", errUpdateInsertFailure, err)
+		}
+	default:
+		return fmt.Errorf("%w [%d]", errInvalidAction, u.Action)
+	}
+	return nil
+}
+
+// updateBidAskByPrice updates the bid and ask spread by supplied updates, this will trim total length of depth level to
+// a specified supplied number
```

Comment:
🐒  `specified supplied number` doesn't seem correct, since we're not supplying it here
I'm not even sure if we need to restate that; since maxDepth is maxDepth and it being enforced everywhere feels implicit
```suggestion
// updateBidAskByPrice updates the bid and ask spread and enforces Depth.options.maxDepth
```
or
```
// updateBidAskByPrice updates the bid and ask spread
```